router/joinfilter.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Stichting ThingsIX Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package router
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/FastFilter/xorfilter"
+	"github.com/ThingsIXFoundation/packet-handling/utils"
+	"github.com/ThingsIXFoundation/router-api/go/router"
+	"github.com/chirpstack/chirpstack/api/go/v4/api"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"time"
+)
+
+type JoinFilterGenerator interface {
+	JoinFilter(ctx context.Context) (*router.JoinFilter, error)
+	UpdateFilter(ctx context.Context) error
+}
+
+func NewJoinFilterGenerator(config RouterConfig) (JoinFilterGenerator, error) {
+	if config.JoinFilterGenerator.ChirpStack.Target != "" {
+		return newChirpstackGenerator(config)
+	}
+	return nil, fmt.Errorf("unknown JoinFilterGenerator")
+}
+
+type chirpstackGenerator struct {
+	dsc    api.DeviceServiceClient
+	filter *xorfilter.Xor8
```

Comment:
I believe this filter should also be protected with a mutex?
It is used in `JoinFilter` and `UpdateFilter`, both called in separate go-routines.